pkg/k8sops/generators.go

```diff
@@ -0,0 +1,307 @@
+// THE SOFTWARE.
+
+package k8sops
+
+import (
+	"fmt"
+
+	m3dboperator "github.com/m3db/m3db-operator/pkg/apis/m3dboperator"
+	myspec "github.com/m3db/m3db-operator/pkg/apis/m3dboperator/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// GenerateCRD generates the crd object needed for the M3DBCluster
+func (k *k8sops) GenerateCRD() *apiextensionsv1beta1.CustomResourceDefinition {
+	return &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: m3dboperator.Name,
+		},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+			Group:   m3dboperator.GroupName,
+			Version: m3dboperator.Version,
+			Scope:   apiextensionsv1beta1.NamespaceScoped,
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural: m3dboperator.ResourcePlural,
+				Kind:   m3dboperator.ResourceKind,
+			},
+		},
+	}
+}
+
+const (
+	_probeTimeoutSeconds      = 30
+	_probeInitialDelaySeconds = 10
+	_probeFailureThreshold    = 15
+	_probePort                = 7201
+	_probePath                = "/health"
+
+	_baseImage              = "quay.io/m3/m3dbnode:latest"
+	_dataDirectory          = "/var/lib/m3db/"
+	_configurationDirectory = "/etc/m3db/"
+	_configurationName      = "m3-configuration"
+	_configurationFileName  = "m3.yml"
+)
+
+var (
+	_configurationFileLocation = fmt.Sprintf("%s%s", _configurationDirectory, _configurationFileName)
+)
+
+// GenerateStatefulSet provides a statefulset object for a m3db cluster
+func (k *k8sops) GenerateStatefulSet(
+	cluster *myspec.M3DBCluster,
+	clusterSpec myspec.ClusterSpec,
+	svcCfg myspec.ServiceConfiguration,
+	isolationGroup string,
+	instanceAmount *int32,
+) (*appsv1.StatefulSet, error) {
+	clusterName := cluster.GetName()
+	ssName := k.StatefulSetName(clusterName, isolationGroup)
+	ports := k.GenerateContainerPorts(svcCfg.Ports)
+
+	// Parse CPU / Memory
+	limitCPU, err := resource.ParseQuantity(clusterSpec.Resources.Limits.CPU)
+	if err != nil {
+		return nil, err
+	}
+	limitMemory, err := resource.ParseQuantity(clusterSpec.Resources.Limits.Memory)
+	if err != nil {
+		return nil, err
+	}
+	requestCPU, err := resource.ParseQuantity(clusterSpec.Resources.Requests.CPU)
+	if err != nil {
+		return nil, err
+	}
+	requestMemory, err := resource.ParseQuantity(clusterSpec.Resources.Requests.Memory)
+	if err != nil {
+		return nil, err
+	}
+
+	probe := &v1.Probe{
+		TimeoutSeconds:      _probeTimeoutSeconds,
+		InitialDelaySeconds: _probeInitialDelaySeconds,
+		FailureThreshold:    _probeFailureThreshold,
+		Handler: v1.Handler{
+			HTTPGet: &v1.HTTPGetAction{
+				Port:   intstr.FromInt(_probePort),
+				Path:   _probePath,
+				Scheme: v1.URISchemeHTTP,
+			},
+		},
+	}
+	statefulset := &appsv1.StatefulSet{
```

Comment:
can you use the same hackery as the other spot to make this a little cleaner?

Reply:
Actually I want to move to another resource type in the next diff. Could we hold off on those changes till then?